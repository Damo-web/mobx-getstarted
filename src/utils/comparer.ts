import { deepEqual } from "../internal"

export interface IEqualsComparer<T> {
    (a: T, b: T): boolean
}

function identityComparer(a: any, b: any): boolean {
    return a === b
}

function structuralComparer(a: any, b: any): boolean {
    return deepEqual(a, b)
}

function defaultComparer(a: any, b: any): boolean {
    return Object.is(a, b)
}

export const comparer = {
    identity: identityComparer,
    structural: structuralComparer,
    default: defaultComparer
}
