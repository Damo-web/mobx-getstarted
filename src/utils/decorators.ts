import { EMPTY_ARRAY, addHiddenProp, fail } from "../internal"

export const mobxDidRunLazyInitializersSymbol = Symbol("mobx did run lazy initializers")
export const mobxPendingDecorators = Symbol("mobx pending decorators")

type DecoratorTarget = {
    [mobxDidRunLazyInitializersSymbol]?: boolean
    [mobxPendingDecorators]?: { [prop: string]: DecoratorInvocationDescription }
}

export type BabelDescriptor = PropertyDescriptor & { initializer?: () => any }

export type PropertyCreator = (
    instance: any,
    propertyName: string,
    descriptor: BabelDescriptor | undefined,
    decoratorTarget: any,
    decoratorArgs: any[]
) => void

type DecoratorInvocationDescription = {
    prop: string
    propertyCreator: PropertyCreator
    descriptor: BabelDescriptor | undefined
    decoratorTarget: any
    decoratorArguments: any[]
}

const enumerableDescriptorCache: { [prop: string]: PropertyDescriptor } = {}
const nonEnumerableDescriptorCache: { [prop: string]: PropertyDescriptor } = {}

function createPropertyInitializerDescriptor(
    prop: string,
    enumerable: boolean
): PropertyDescriptor {
    const cache = enumerable ? enumerableDescriptorCache : nonEnumerableDescriptorCache
    return (
        cache[prop] ||
        (cache[prop] = {
            configurable: true,
            enumerable: enumerable,
            get() {
                initializeInstance(this)
                return this[prop]
            },
            set(value) {
                initializeInstance(this)
                this[prop] = value
            }
        })
    )
}

export function initializeInstance(target: any)
export function initializeInstance(target: DecoratorTarget) {
    if (target[mobxDidRunLazyInitializersSymbol] === true) return
    const decorators = target[mobxPendingDecorators]
    if (decorators) {
        addHiddenProp(target, mobxDidRunLazyInitializersSymbol, true)
        for (let key in decorators) {
            const d = decorators[key]
            d.propertyCreator(target, d.prop, d.descriptor, d.decoratorTarget, d.decoratorArguments)
        }
    }
}

export function createPropDecorator(
    propertyInitiallyEnumerable: boolean,
    propertyCreator: PropertyCreator
) {
    return function decoratorFactory() {
        let decoratorArguments: any[]

        const decorator = function decorate(
            target: DecoratorTarget,
            prop: string,
            descriptor: BabelDescriptor | undefined,
            applyImmediately?: any
            // This is a special parameter to signal the direct application of a decorator, allow extendObservable to skip the entire type decoration part,
            // as the instance to apply the decorator to equals the target
        ) {
            if (applyImmediately === true) {
                propertyCreator(target, prop, descriptor, target, decoratorArguments)
                return null
            }
            if (process.env.NODE_ENV !== "production" && !quacksLikeADecorator(arguments))
                fail("This function is a decorator, but it wasn't invoked like a decorator")
            if (!Object.prototype.hasOwnProperty.call(target, mobxPendingDecorators)) {
                const inheritedDecorators = target[mobxPendingDecorators]
                addHiddenProp(target, mobxPendingDecorators, { ...inheritedDecorators })
            }
            target[mobxPendingDecorators]![prop] = {
                prop,
                propertyCreator,
                descriptor,
                decoratorTarget: target,
                decoratorArguments
            }
            return createPropertyInitializerDescriptor(prop, propertyInitiallyEnumerable)
        }

        if (quacksLikeADecorator(arguments)) {
            // @decorator
            decoratorArguments = EMPTY_ARRAY
            return decorator.apply(null, arguments as any)
        } else {
            // @decorator(args)
            decoratorArguments = Array.prototype.slice.call(arguments)
            return decorator
        }
    } as Function
}

export function quacksLikeADecorator(args: IArguments): boolean {
    return (
        ((args.length === 2 || args.length === 3) && typeof args[1] === "string") ||
        (args.length === 4 && args[3] === true)
    )
}
