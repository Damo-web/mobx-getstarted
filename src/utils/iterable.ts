export function makeIterable<T>(iterator: Iterator<T>): IterableIterator<T> {
    iterator[Symbol.iterator] = self
    return iterator as any
}

function self() {
    return this
}
