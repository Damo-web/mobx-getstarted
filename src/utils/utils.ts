import {
    IKeyValueMap,
    IObservableArray,
    ObservableMap,
    globalState,
    isObservableArray,
    isObservableMap
} from "../internal"

export const OBFUSCATED_ERROR =
    "An invariant failed, however the error is obfuscated because this is an production build."

export const EMPTY_ARRAY = []
Object.freeze(EMPTY_ARRAY)

export const EMPTY_OBJECT = {}
Object.freeze(EMPTY_OBJECT)

export interface Lambda {
    (): void
    name?: string
}

export function getNextId() {
    return ++globalState.mobxGuid
}

export function fail(message: string | boolean): never {
    invariant(false, message)
    throw "X" // unreachable
}

export function invariant(check: false, message?: string | boolean): never
export function invariant(check: true, message?: string | boolean): void
export function invariant(check: any, message?: string | boolean): void
export function invariant(check: boolean, message?: string | boolean) {
    if (!check) throw new Error("[mobx] " + (message || OBFUSCATED_ERROR))
}

/**
 * Prints a deprecation message, but only one time.
 * Returns false if the deprecated message was already printed before
 */
const deprecatedMessages: string[] = []

export function deprecated(msg: string): boolean
export function deprecated(thing: string, replacement: string): boolean
export function deprecated(msg: string, thing?: string): boolean {
    if (process.env.NODE_ENV === "production") return false
    if (thing) {
        return deprecated(`'${msg}', use '${thing}' instead.`)
    }
    if (deprecatedMessages.indexOf(msg) !== -1) return false
    deprecatedMessages.push(msg)
    console.error("[mobx] Deprecated: " + msg)
    return true
}

/**
 * Makes sure that the provided function is invoked at most once.
 */
export function once(func: Lambda): Lambda {
    let invoked = false
    return function() {
        if (invoked) return
        invoked = true
        return (func as any).apply(this, arguments)
    }
}

export const noop = () => {}

export function unique<T>(list: T[]): T[] {
    const res: T[] = []
    list.forEach(item => {
        if (res.indexOf(item) === -1) res.push(item)
    })
    return res
}

export function isObject(value: any): boolean {
    return value !== null && typeof value === "object"
}

export function isPlainObject(value) {
    if (value === null || typeof value !== "object") return false
    const proto = Object.getPrototypeOf(value)
    return proto === Object.prototype || proto === null
}

export function makeNonEnumerable(object: any, propNames: PropertyKey[]) {
    for (let i = 0; i < propNames.length; i++) {
        addHiddenProp(object, propNames[i], object[propNames[i]])
    }
}

export function addHiddenProp(object: any, propName: PropertyKey, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: true,
        configurable: true,
        value
    })
}

export function addHiddenFinalProp(object: any, propName: PropertyKey, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: false,
        configurable: true,
        value
    })
}

export function isPropertyConfigurable(object: any, prop: PropertyKey): boolean {
    const descriptor = Object.getOwnPropertyDescriptor(object, prop)
    return !descriptor || (descriptor.configurable !== false && descriptor.writable !== false)
}

export function assertPropertyConfigurable(object: any, prop: PropertyKey) {
    if (process.env.NODE_ENV !== "production" && !isPropertyConfigurable(object, prop))
        fail(
            `Cannot make property '${prop.toString()}' observable, it is not configurable and writable in the target object`
        )
}

export function createInstanceofPredicate<T>(
    name: string,
    clazz: new (...args: any[]) => T
): (x: any) => x is T {
    const propName = "isMobX" + name
    clazz.prototype[propName] = true
    return function(x) {
        return isObject(x) && x[propName] === true
    } as any
}

/**
 * Returns whether the argument is an array, disregarding observability.
 */
export function isArrayLike(x: any): x is Array<any> | IObservableArray<any> {
    return Array.isArray(x) || isObservableArray(x)
}

export function isES6Map(thing): boolean {
    return thing instanceof Map
}

export function getMapLikeKeys<K, V>(map: ObservableMap<K, V>): ReadonlyArray<K>
export function getMapLikeKeys<V>(map: IKeyValueMap<V> | any): ReadonlyArray<string>
export function getMapLikeKeys(map: any): any {
    if (isPlainObject(map)) return Object.keys(map)
    if (Array.isArray(map)) return map.map(([key]) => key)
    if (isES6Map(map) || isObservableMap(map)) return Array.from(map.keys())
    return fail(`Cannot get keys from '${map}'`)
}

export function toPrimitive(value) {
    return value === null ? null : typeof value === "object" ? "" + value : value
}
