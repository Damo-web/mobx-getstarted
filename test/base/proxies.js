import {
    computed,
    decorate,
    isComputedProp,
    isAction,
    isObservableProp,
    autorun,
    observable,
    action,
    reaction,
    extendObservable,
    keys,
    $mobx
} from "../../src/mobx.ts"

import { stripAdminFromDescriptors } from "../utils/test-utils"

test("should react to key removal (unless reconfiguraing to empty) - 1", () => {
    const events = []
    const x = observable.object({
        y: 1,
        z: 1
    })

    reaction(() => Object.keys(x), keys => events.push(keys.join(",")), { fireImmediately: true })
    expect(events).toEqual(["y,z"])
    delete x.y
    expect(events).toEqual(["y,z", "z"])
    // should not trigger another time..
    delete x.y
    expect(events).toEqual(["y,z", "z"])
})

test("should react to key removal (unless reconfiguraing to empty) - 2", () => {
    const events = []
    const x = observable.object({
        y: 1,
        z: 1
    })

    reaction(() => x.z, v => events.push(v))

    delete x.z
    expect(events).toEqual([undefined])
})

test("should react to key removal (unless reconfiguraing to empty) - 2", () => {
    const events = []
    const x = observable.object({
        y: 1,
        z: undefined
    })

    reaction(() => x.z, v => events.push(v))

    delete x.z
    expect(events).toEqual([])
})

test("should react to future key additions - 1", () => {
    const events = []
    const x = observable.object({})

    reaction(() => Object.keys(x), keys => events.push(keys.join(",")))

    x.y = undefined
    expect(events).toEqual(["y"])
})

test("should react to future key additions - 2", () => {
    const events = []
    const x = observable.object({})

    reaction(
        () => {
            return x.z
        },
        v => {
            events.push(v)
        }
    )

    x.z = undefined
    expect(Object.keys(x)).toEqual(["z"])
    x.y = 3
    expect(events).toEqual([])
    delete x.y
    expect(events).toEqual([])
    x.z = 4
    expect(events).toEqual([4])
})

test("should throw clear warning if trying to add computed to already reserved key", () => {
    const x = observable.object({ z: 3 })

    expect(() => {
        extendObservable(x, {
            get z() {
                return 3
            }
        })
    }).toThrow(/can only be used to introduce new properties/)
})

test("correct keys are reported", () => {
    const x = observable.object({
        x: 1,
        get y() {
            return 2
        }
    })
    x.z = 3
    extendObservable(x, {
        a: 4,
        get b() {
            return 5
        }
    })
    x.y
    x.b // make sure it is read

    expect(Object.keys(x)).toEqual(["x", "z", "a"])
    expect(Object.values(x)).toEqual([1, 3, 4])
    expect(Object.entries(x)).toEqual([["x", 1], ["z", 3], ["a", 4]])

    expect(Object.getOwnPropertyNames(x)).toEqual(["x", "y", "z", "a", "b"])
    expect(keys(x)).toEqual(["x", "z", "a"])

    delete x.x
    expect(Object.keys(x)).toEqual(["z", "a"])
    expect(Object.getOwnPropertyNames(x)).toEqual(["y", "z", "a", "b"])
    expect(keys(x)).toEqual(["z", "a"])
})

test("in operator", () => {
    const x = observable.object({
        x: 1,
        get y() {
            return 2
        }
    })
    x.z = 3
    extendObservable(x, {
        a: 4,
        get b() {
            return 5
        }
    })
    expect("x" in x).toBeTruthy()
    expect("y" in x).toBeTruthy()
    expect("a" in x).toBeTruthy()
    expect("b" in x).toBeTruthy()
    expect("z" in x).toBeTruthy()
    expect("c" in x).toBeFalsy()
    expect("c" in x).toBeFalsy() // not accidentally create
    delete x.x
    expect("x" in x).toBeFalsy()
})

test("for-in operator", () => {
    const x = observable.object({
        x: 1,
        get y() {
            return 2
        }
    })
    x.z = 3
    extendObservable(x, {
        a: 4,
        get b() {
            return 5
        }
    })

    function computeKeys() {
        const res = []
        for (let key in x) res.push(key)
        return res
    }

    expect(computeKeys(x)).toEqual(["x", "z", "a"])
    delete x.x
    expect(computeKeys(x)).toEqual(["z", "a"])
})

test("type coercion doesn't break", () => {
    const x = observable({})
    expect("" + x).toBe("[object Object]")
    expect(42 * x).toBeNaN()
})

test("adding a different key doesn't trigger a pending key", () => {
    const x = observable({})
    let counter = 0

    const d = autorun(() => {
        x.x
        counter++
    })
    expect(counter).toBe(1)

    x.y = 3
    expect(counter).toBe(1)

    x.x = 3
    expect(counter).toBe(2)

    d()
})

test("proxy false reverts to original behavior", () => {
    const x = observable({ x: 3 }, {}, { proxy: false })
    x.y = 3
    expect(isObservableProp(x, "x")).toBe(true)
    expect(isObservableProp(x, "y")).toBe(false)
})

test("ownKeys invariant not broken - 1", () => {
    const a = observable({ x: 3, get y() {} })
    expect(() => {
        Object.freeze(a)
    }).toThrow("cannot be frozen")
})

test("ownKeys invariant not broken - 2", () => {
    const a = observable([2])
    expect(() => {
        Object.freeze(a)
    }).toThrow("cannot be frozen")
})

test("non-proxied object", () => {
    const a = observable({ x: 3 }, {}, { proxy: false })
    a.b = 4
    debugger
    extendObservable(
        a,
        {
            double() {
                this.x = this.x * 2
            },
            get y() {
                return this.x * 2
            }
        },
        {
            double: action
        }
    )

    expect(a.y).toBe(6)
    debugger
    a.double()
    expect(a.y).toBe(12)
    expect(isComputedProp(a, "y")).toBe(true)
    expect(isAction(a.double)).toBe(true)
    expect(stripAdminFromDescriptors(Object.getOwnPropertyDescriptors(a))).toMatchSnapshot()
    expect(Object.keys(a)).toEqual(["x", "b"])
})

test("extend proxies", () => {
    const a = observable({ x: 3 })
    a.b = 4
    extendObservable(
        a,
        {
            double() {
                this.x = this.x * 2
            },
            get y() {
                return this.x * 2
            }
        },
        {
            double: action
        }
    )

    expect(a.y).toBe(6)
    debugger
    a.double()
    expect(a.y).toBe(12)
    expect(isComputedProp(a, "y")).toBe(true)
    expect(isAction(a.double)).toBe(true)
    expect(stripAdminFromDescriptors(Object.getOwnPropertyDescriptors(a))).toMatchSnapshot()
    expect(Object.keys(a)).toEqual(["x", "b"])
})

test("decorate proxies", () => {
    const a = observable({ x: 3 })
    a.b = 4
    extendObservable(
        a,
        {
            double() {
                this.x = this.x * 2
            },
            get y() {
                return this.x * 2
            }
        },
        {
            double: action
        }
    )

    expect(a.y).toBe(6)
    a.double()
    expect(a.y).toBe(12)
    expect(isComputedProp(a, "y")).toBe(true)
    expect(isAction(a.double)).toBe(true)
    expect(stripAdminFromDescriptors(Object.getOwnPropertyDescriptors(a))).toMatchSnapshot()
    expect(Object.keys(a)).toEqual(["x", "b"])
})

test("predictable 'this' - 1", () => {
    debugger
    const a = observable.object(
        {
            a0() {
                return this
            },
            a1() {
                return this
            },
            a2() {
                return this
            },
            get computed() {
                return this
            }
        },
        {
            a1: action,
            a2: action.bound
        }
    )

    expect(a.a0()).toBe(a)
    expect(a.a1()).toBe(a)
    expect(a.a2()).toBe(a) // pre-bound!
    expect(a.computed).toBe(a)
})

test("predictable 'this' - 2", () => {
    class A {
        a0() {
            return this
        }

        @action
        a1() {
            return this
        }

        @action.bound
        a2() {
            return this
        }

        @computed
        get computed() {
            return this
        }
    }

    const a = new A()

    expect(a.a0()).toBe(a)
    expect(a.a1()).toBe(a)
    expect(a.a2()).toBe(a)
    expect(a.computed).toBe(a)
})

test("1796 - deleting / recreate prop", () => {
    let value = observable({
        foo: undefined // if foo is something like 'abc', it works.
    })

    const events = []

    const aut = autorun(() => {
        events.push(value.foo)
    })
    delete value.foo
    value.foo = "def"
    expect(events).toEqual([
        undefined,
        undefined, // ideally  not, but ok..
        "def"
    ])
})
