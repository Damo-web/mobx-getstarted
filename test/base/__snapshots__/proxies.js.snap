// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorate proxies 1`] = `
Object {
  "b": Object {
    "configurable": true,
    "enumerable": true,
    "get": [Function],
    "set": [Function],
  },
  "double": Object {
    "configurable": true,
    "enumerable": false,
    "value": [Function],
    "writable": true,
  },
  "x": Object {
    "configurable": true,
    "enumerable": true,
    "get": [Function],
    "set": [Function],
  },
  "y": Object {
    "configurable": true,
    "enumerable": false,
    "get": [Function],
    "set": [Function],
  },
  Symbol(mobx administration): Object {
    "configurable": true,
    "enumerable": false,
    "value": "(omitted)",
    "writable": true,
  },
}
`;

exports[`extend proxies 1`] = `
Object {
  "b": Object {
    "configurable": true,
    "enumerable": true,
    "get": [Function],
    "set": [Function],
  },
  "double": Object {
    "configurable": true,
    "enumerable": false,
    "value": [Function],
    "writable": true,
  },
  "x": Object {
    "configurable": true,
    "enumerable": true,
    "get": [Function],
    "set": [Function],
  },
  "y": Object {
    "configurable": true,
    "enumerable": false,
    "get": [Function],
    "set": [Function],
  },
  Symbol(mobx administration): Object {
    "configurable": true,
    "enumerable": false,
    "value": "(omitted)",
    "writable": true,
  },
}
`;

exports[`non-proxied object 1`] = `
Object {
  "b": Object {
    "configurable": true,
    "enumerable": true,
    "value": 4,
    "writable": true,
  },
  "double": Object {
    "configurable": true,
    "enumerable": false,
    "value": [Function],
    "writable": true,
  },
  "x": Object {
    "configurable": true,
    "enumerable": true,
    "get": [Function],
    "set": [Function],
  },
  "y": Object {
    "configurable": true,
    "enumerable": false,
    "get": [Function],
    "set": [Function],
  },
  Symbol(mobx administration): Object {
    "configurable": true,
    "enumerable": false,
    "value": "(omitted)",
    "writable": true,
  },
}
`;
